File: 0198. House Robber.java
// Medium
// DP
// Time:O(n), Space:O(n)
// https://leetcode.cn/problems/house-robber/

class Solution {
  // dp: means the max money can rob from 0 to i
  // if rob the ith house, the max is dp[i - 2] + nums[i]
  // if not rob the ith house, the max is dp[i - 1](from prev)
  public int rob(int[] nums) {
      // basecase: if (nums.length == 0 || nums == null) return 0;
      if (nums.length == 1) return nums[0];
      int[] dp = new int[nums.length + 1];
      dp[0] = nums[0];
      dp[1] = Math.max(dp[0], nums[1]);// rob cur or prev(dp[0])
      
      for (int i = 2; i < nums.length; i++) {
          dp[i] = Math.max(dp[i - 1], dp[i - 2] + nums[i]);
      }
      return dp[nums.length - 1]; // from 0 to n-1
  }
}

File: 0198. House Robber.py

from typing import *

# two options: rob cur(dp[i-2] + nums[i]) or not(keep dp[i - 1])
# Time:O(n), Space:O(1)

class Solution:
    def rob(self, nums: List[int]) -> int:
        # base
        if not nums:
            return 0
        
        if len(nums) == 1:
            return nums[0]
        
        prev2 = nums[0]
        prev1 = max(nums[0], nums[1])

        for i in range(2, len(nums)):
            cur = max(prev1, prev2 + nums[i])
            prev2 = prev1
            prev1 = cur
        
        return prev1  # last one dp[n-1]  
        